Enforce vscode and ads version check when installing extensions

alanrenmsft opened the pull request:
#4250 
Problem:
Right now we don't have a way to stop a user from installing an extension that is not compatible with ADS.
include the following scenarios: 
1. Install new extension from ADS marketplace
2. Update/Auto update of installed extension when there is a new version released
3. Install new extension using VSIX file

Solution:
Enforce the version check for the above scenarios.
Behavior with the fix:
1. install new extension from marketplace or vsix file:
![image](https://user-images.githubusercontent.com/13777222/53702865-72c46700-3dc0-11e9-86ed-c362871a86f5.png)

2. autoupdate of installed extension
where there is an update to the installed extension, but is not compatible with the current ADS, the local extension won't be auto updated, instead, the text of the update button will appear like 'Update to xxxx'. 
![image](https://user-images.githubusercontent.com/13777222/53702861-64764b00-3dc0-11e9-8d60-0fb905641000.png)

when user clicks the update button, the similar error message will be prompted indicating the incompatibility.

Notes:
Why am I not using the vscode in the engine and repurpose it for the ADS version check? 
we will have extensions built for vscode supported in ADS, we need to keep the vscode version for this scenario.

anthonydresser commented on src/vs/platform/extensionManagement/common/extensionManagement.ts:
```diff
@@ -141,6 +141,8 @@ export interface IGalleryExtensionProperties {
 	dependencies?: string[];
 	extensionPack?: string[];
 	engine?: string;
```
does renaming these `vscodeEngine` and `azdataEngine` make more sense for readability?

alanrenmsft replied:
I can rename it to azdataEngine, but we should keep the engine for vscode. as I mentioned in the PR description, we need to make sure the vscode extensions we would like to bring over to ADS also can work without modifications.

anthonydresser replied:
I mean to make sure vscode extensions work with ads all we have to do is not change the property we read from package.json. We can also change the interfaces and what we call properties we our code since that won't affect what we read from the extensions.

anthonydresser commented on src/vs/platform/extensionManagement/node/extensionGalleryService.ts:
```diff
@@ -123,6 +123,8 @@ const PropertyType = {
 	Dependency: 'Microsoft.VisualStudio.Code.ExtensionDependencies',
 	ExtensionPack: 'Microsoft.VisualStudio.Code.ExtensionPack',
 	Engine: 'Microsoft.VisualStudio.Code.Engine',
```
Same here, does it make more sense to use `VscodeEngine` and `AzdataEngine`

alanrenmsft replied:
same as above, will keep the engine and rename to AzDataEngine

anthonydresser commented on src/vs/platform/extensionManagement/node/extensionGalleryService.ts:
```diff
@@ -724,7 +734,8 @@ export class ExtensionGalleryService implements IExtensionGalleryService {
 	}
 
 	loadCompatibleVersion(extension: IGalleryExtension, fromVersion: string = extension.version): Promise<IGalleryExtension> {
-		if (extension.version === fromVersion && extension.properties.engine && isEngineValid(extension.properties.engine)) {
```
Whats the purpose of disabling the additional engine checks?

alanrenmsft replied:
added comments explaining why.

anthonydresser commented on src/vs/workbench/parts/extensions/electron-browser/extensionsActions.ts:
```diff
@@ -191,6 +191,11 @@ export class InstallAction extends Action {
 				return this.notificationService.error(err);
 			}
 
+			// {{SQL CARBON EDIT}}
```
Why do we need changes here?

alanrenmsft replied:
added comments

anthonydresser replied:
This is already handled later by vscode;
https://github.com/Microsoft/azuredatastudio/pull/4267/files#diff-148a21c9158278aadbd155843ed47c36L64

anthonydresser commented on src/vs/workbench/parts/extensions/electron-browser/extensionsActions.ts:
```diff
@@ -418,6 +423,11 @@ export class UpdateAction extends Action {
 			}
 
+			// {{SQL CARBON EDIT}}
+			if(err && err.code === INSTALL_ERROR_INCOMPATIBLE) {
```
Why do we need changes here?

alanrenmsft replied:
added comments

anthonydresser replied:
This is already handled later by vscode;
https://github.com/Microsoft/azuredatastudio/pull/4267/files#diff-148a21c9158278aadbd155843ed47c36L64

anthonydresser commented on src/vs/workbench/parts/extensions/node/extensionsWorkbenchService.ts:
```diff
@@ -703,6 +709,11 @@ export class ExtensionsWorkbenchService implements IExtensionsWorkbenchService,
 				}
+			}
+			// {{SQL CARBON EDIT}}
+			, (error) => {
```
Why do we need changes here?

alanrenmsft replied:
added comments

anthonydresser replied:
I would presume vscode handles if a vsix extension install fails somewhere? They really don't handle that?

anthonydresser commented on src/vs/workbench/parts/extensions/node/extensionsWorkbenchService.ts:
```diff
@@ -721,6 +732,12 @@ export class ExtensionsWorkbenchService implements IExtensionsWorkbenchService,
 		}
 
+		// {{SQL CARBON EDIT}}
+		if (gallery.properties.engine && (!isEngineValid(gallery.properties.engine, product.vscodeVersion)
```
Why do we need changes here?

alanrenmsft replied:
added comments

alanrenmsft commented:
@kburtram , in the latest commit, I added telemetry for install from vsix scenario as it is a very common scenario for ADS (all external extensions are installed this way). we already have install from gallery telemetry enabled, I just added the version property which I think is an pretty important information.

kevcunnane commented:
@alanrenmsft - for new telemetry can you sync with @yualan as this may need privacy review before release.

alanrenmsft commented:
@kevcunnane Technically this is not new telemetry, previously we only have telemetry for install extension from gallery scenario. I am enabling telemetry for the install from vsix scenario. the only new information is the extension's version. which I think should be fine. But thanks for reminding, I will let @yualan to decide whether it needs privacy review.

anthonydresser commented:
I had some comments that haven't been addressed, can you address them in a new PR. I think you have unnecessary changes in vscode source.

alanrenmsft commented:
@anthonydresser sure, thanks!
